protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,713 @@
+        self.ongoing_outbound.is_some()
+    }
+
+    pub fn reachability(&self) -> Reachability {
+        self.reachability.clone()
+    }
+
+    pub fn ongoing_probe(&self) -> Option<&Probe> {
+        self.ongoing_outbound.as_ref()
+    }
+
+    pub fn add_server(&mut self, peer: PeerId, address: Option<Multiaddr>) {
+        self.static_servers.push(peer);
+        if let Some(addr) = address {
+            self.inner.add_address(&peer, addr);
+        }
+    }
+
+    pub fn remove_server(&mut self, peer: &PeerId) {
+        self.static_servers.retain(|p| p != peer)
+    }
+
+    fn new_probe(
+        &self,
+        required_success: usize,
+        mut servers: Vec<PeerId>,
+        extend_with_static: bool,
+        extend_with_connected: bool,
+        max_peers: usize,
+        min_valid: usize,
+    ) -> Probe {
+        servers.truncate(max_peers);
+        if servers.len() < max_peers && extend_with_static {
+            servers.extend(self.static_servers.clone())
+        }
+        if servers.len() < max_peers && extend_with_connected {
+            let mut connected = self.connected.iter();
+            // TODO: use random set
+            for _ in 0..connected.len() {
+                let peer = connected.next().unwrap();
+                servers.push(*peer);
+                if servers.len() >= max_peers {
+                    break;
+                }
+            }
+        }
+        Probe {
+            server_count: servers.len(),
+            pending_servers: servers,
+            addresses: HashMap::new(),
+            required_success,
+            dismissed: 0,
+            errors: Vec::new(),
+            min_valid,
+        }
+    }
+
+    fn do_probe(&mut self, probe: Probe, addr_records: Vec<AddressRecord>) {
+        let addrs: Vec<Multiaddr> = addr_records.into_iter().map(|r| r.addr).collect();
+        for peer_id in probe.pending_servers.clone() {
+            self.inner.send_request(
+                &peer_id,
+                DialRequest {
+                    peer_id,
+                    addrs: addrs.clone(),
+                },
+            );
+        }
+        let _ = self.ongoing_outbound.insert(probe);
+    }
+
+    // Handle the inbound request and collect the valid addresses to be dialed.
+    fn handle_request(&mut self, sender: PeerId, request: DialRequest) -> Option<Vec<Multiaddr>> {
+        let config = self
+            .server_config
+            .as_ref()
+            .expect("Server config is present.");
+
+        // Validate that the peer to be dialed is the request's sender.
+        if request.peer_id != sender {
+            return None;
+        }
+        // Check that there is no ongoing dial to the remote.
+        if self.ongoing_inbound.contains_key(&sender) {
+            return None;
+        }
+        // Check if max simultaneous autonat dial-requests are reached.
+        if self.ongoing_inbound.len() >= config.max_ongoing {
+            return None;
+        }
```

Comment:
These might deserve a `log::trace` or `log::debug` to not be completely silent about what is happening.

Reply:
Added in b70428e28c4d2a2315598deeb7ebd43838473ac8.